spec/lib/utils/job-utils/http-tool-spec.js

```diff
@@ -0,0 +1,178 @@
+            ])
+        );
+        var HttpTool = helper.injector.get('JobUtils.HttpTool');
+        httpTool = new HttpTool();
+    });
+
+    it('Should handle basic auth', function(done){
+        nock(siteGen).get('/basicAuth')
+        .basicAuth({user:'hello', pass:'world'})
+        .reply(200);
+
+        requestSettings.url = siteGen + '/basicAuth';
+        requestSettings.method = 'GET';
+        requestSettings.credential = {username:'hello', password:'world'};
+        // requestSettings.data = '{hello: world}';
+
+        httpTool.setupRequest(requestSettings)
+        .then(httpTool.runRequest()
+        .then(function(data){
+            expect(data).to.have.property('httpStatusCode').to.equal(200);
+            done();
+        }));
+    });
+
+    it('can handle header change', function(done){
+        nock(siteGen)
+        .matchHeader('cookie', 'mySession')
+        .get('/withCookie')
+        .reply(200);
+
+        requestSettings.url = siteGen + '/withCookie';
+        requestSettings.method = 'GET';
+        requestSettings.headers = {cookie: 'mySession'};
+        
+        httpTool.setupRequest(requestSettings)
+        .then(httpTool.runRequest()
+        .then(function(data){
+            expect(data).to.have.property('httpStatusCode').to.equal(200);
+            done();
+        }));
+    });
+
+    it('can handle secure http and non-standard port', function(done){
+        nock('https://mysite.emc.com:12345')
+        .get('/non-standard-port/http-secure')
+        .reply(200, 'You are good');
+        
+        requestSettings.url = 'https://mysite.emc.com:12345/non-standard-port/http-secure';
+        requestSettings.method = 'GET';
+        
+        httpTool.setupRequest(requestSettings)
+        .then(httpTool.runRequest()
+        .then(function(data){
+            expect(data).to.have.property('body').to.equal('You are good');
+            done();
+        }));
+    });
+
+    it('can handle object formatted url', function(done){
+        nock(siteGen).get('/getWithObjUrl').reply(200);
+
+        requestSettings.url = {
+            protocol: "http",
+            host: "mysite.emc.com",
+            path: "/getWithObjUrl"
+        };
+        requestSettings.method = 'GET';
+
+        httpTool.setupRequest(requestSettings)
+        .then(httpTool.runRequest()
+        .then(function(data){
+            expect(data).to.have.property('httpStatusCode').to.equal(200); 
+            done();
+        }));
+    });
+    
+    it('can put err into reject callback', function(done){
+        nock(siteGen).get('/good-get').reply(200);
+
+        requestSettings.url = siteGen + '/bad-get';
+        requestSettings.method = 'GET';
+        
+        httpTool.setupRequest(requestSettings)
+        .then(httpTool.runRequest()
+        .catch(function(err){
+            expect(err).to.have.property('status').to.equal(404);
+            done();
+        }));
```

Comment:
since you are expecting the promise is rejected, so you need to explicitly fail the test case if the promise is fulfilled.
```javascript
return httpTool.setupRequest(requestSettings)
         .then(function() {
             return httpTool.runRequest();
         })
         .then(function() {
             done(new Error('Should never run here, otherwise the code has some bug.'));
         });
         .catch(function(err){
             expect(err).to.have.property('status').to.equal(404);
             done();
         }));
```